File: get_recent_tracks.R
library(dplyr)
library(spotifyr)
library(RSQLite)

source("api_keys.R")

source("connect_to_db.R")

spotify_auth_code = get_spotify_authorization_code(client_id = spotify_client_id, client_secret = spotify_client_secret, scope = "user-read-recently-played")


query = paste("SELECT MAX(timestamp) FROM api_tracks_played")

latest_timestamp = unlist(dbGetQuery(db, query))

get_recent_tracks = function(authorization = spotify_auth_code) {
  response = NULL
  try(response <- get_my_recently_played(limit = 50, authorization = authorization, after = latest_timestamp))
  if (is.data.frame(response)) {
    recent_tracks = response %>% 
      select(timestamp = played_at, spotify_id = track.id, track_name = track.name)
    
    dbAppendTable(db, "api_tracks_played", recent_tracks)
    
    latest_timestamp <<- max(recent_tracks$timestamp)
  }
  
  later::later(get_recent_tracks, 300)
  
  message(paste0(lubridate::now("UTC"), ": last recorded timestamp: ", latest_timestamp))
}

get_recent_tracks()

File: get_recent_tracks.py
from spotipy.oauth2 import SpotifyOAuth
import spotipy
import sqlite3
import yaml
import logging
import logging.handlers

# start logger
log_format = '%(asctime)s %(levelname)s: %(message)s'
log_date_format = 'spotify-logger:'
handler = [logging.handlers.SysLogHandler(address='/dev/log')]
logging.basicConfig(handlers=handler, level="INFO",
                    format=log_format, datefmt=log_date_format)

# read in api keys
with open('api_keys.yaml') as file:
    keys = yaml.full_load(file)

# connect to the spotify api
try:
    sp = spotipy.Spotify(auth_manager=SpotifyOAuth(scope='user-read-recently-played',
                                                   client_secret=keys['spotify']['client secret'],
                                                   client_id=keys['spotify']['client id'],
                                                   redirect_uri='http://localhost:1410/'))
except Exception as e:
    logging.exception("Could not connect to Spotify.")

# connect to the database
try:
    conn = sqlite3.connect(keys['db location'])
    db = conn.cursor()
except Exception as e:
    logging.exception("Could not connect to database.")

# get the timestamp of the latest record from the database
last_track_query = 'SELECT MAX(timestamp) FROM scrobbles;'
db.execute(last_track_query)
last_track_ts = db.fetchone()

# poll spotify for tracks played since the latest record in the database
try:
    results = sp.current_user_recently_played(after=last_track_ts)
except Exception as e:
    logging.exception("Error fetching recent tracks from Spotify.")

# prep results for writing to database
def parse_play_record(record):
    timestamp = record['played_at']
    spotify_id = record['track']['id']
    track_name = record['track']['name']

    # return a tuple that can be passed to the database
    return((timestamp, spotify_id, track_name))


# combine all of the new records into a list
try:
    tracks = [parse_play_record(i) for i in results['items']]
except Exception as e:
    logging.exception("Error parsing tracks.")

# write the new list to the database
try:
    insert_query = "INSERT INTO scrobbles ('timestamp', 'spotify_id', 'track_name') VALUES (?, ? ,?);"
    # execute the insert query
    db.executemany(insert_query, tracks)
    # write the result to the database
    conn.commit()
except Exception as e:
    logging.exception("Error writing to database.")

# disconnect from the database
conn.close()

# log how many tracks were recorded
logging.info("Scrobbled %d tracks.", len(tracks))
